src/client/csfle/client_encryption.rs

```diff
@@ -0,0 +1,354 @@
+    pub key_alt_names: Option<Vec<String>>,
+    /// An optional buffer of 96 bytes to use as custom key material for the data key being
+    /// created.  If unset, key material for the new data key is generated from a cryptographically
+    /// secure random device.
+    #[builder(default)]
+    pub key_material: Option<Vec<u8>>,
+}
+
+/// A KMS-specific key used to encrypt data keys.
+#[derive(Serialize)]
+#[serde(rename_all = "camelCase", untagged)]
+#[non_exhaustive]
+#[allow(missing_docs)]
+pub enum MasterKey {
+    Aws {
+        region: String,
+        /// The Amazon Resource Name (ARN) to the AWS customer master key (CMK).
+        key: String,
+        /// An alternate host identifier to send KMS requests to. May include port number. Defaults
+        /// to "kms.<region>.amazonaws.com"
+        endpoint: Option<String>,
+    },
+    Azure {
+        /// Host with optional port. Example: "example.vault.azure.net".
+        key_vault_endpoint: String,
+        key_name: String,
+        /// A specific version of the named key, defaults to using the key's primary version.
+        key_version: Option<String>,
+    },
+    Gcp {
+        project_id: String,
+        location: String,
+        key_ring: String,
+        key_name: String,
+        /// A specific version of the named key, defaults to using the key's primary version.
+        key_version: Option<String>,
+        /// Host with optional port. Defaults to "cloudkms.googleapis.com".
+        endpoint: Option<String>,
+    },
+    /// Master keys are not applicable to `KmsProvider::Local`.
+    Local,
+    Kmip {
+        /// keyId is the KMIP Unique Identifier to a 96 byte KMIP Secret Data managed object.  If
+        /// keyId is omitted, the driver creates a random 96 byte KMIP Secret Data managed object.
+        key_id: Option<String>,
+        /// Host with optional port.
+        endpoint: Option<String>,
+    },
+}
+
+// #[non_exhaustive]
+// pub struct RewrapManyDataKeyOptions {
+// pub provider: KmsProvider,
+// pub master_key: Option<Document>,
+// }
+//
+//
+// #[non_exhaustive]
+// pub struct RewrapManyDataKeyResult {
+// pub bulk_write_result: Option<BulkWriteResult>,
+// }
+
+/// The options for explicit encryption.
+#[derive(TypedBuilder)]
+#[builder(field_defaults(setter(into)))]
+#[non_exhaustive]
+pub struct EncryptOptions {
+    /// The key to use.
+    pub key: EncryptKey,
+    /// The encryption algorithm.
+    ///
+    /// To insert or query with an "Indexed" encrypted payload, use a `Client` configured with
+    /// `AutoEncryptionOptions`. `AutoEncryptionOptions.bypass_query_analysis may be true.
+    /// `AutoEncryptionOptions.bypass_auto_encryption` must be false.
+    pub algorithm: Algorithm,
+    /// The contention factor.
+    #[builder(default)]
+    pub contention_factor: Option<i64>,
+    /// The query type.
+    #[builder(default)]
+    pub query_type: Option<String>,
+}
+
+/// An encryption key reference.
+pub enum EncryptKey {
```

Comment:
Do we need `#[non_exhaustive]` here? Also, we'll probably want `Debug` and `Clone`.

Reply:
Done.